shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -1363,12 +1363,30 @@ - (void)onLocaleUpdated:(NSNotification*)notification {
 - (void)waitForFirstFrame:(NSTimeInterval)timeout
                  callback:(void (^_Nonnull)(BOOL didTimeout))callback {
   dispatch_queue_t queue = dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0);
+  __weak FlutterEngine* weakSelf = self;
   dispatch_async(queue, ^{
+    FlutterEngine* strongSelf = weakSelf;
+    if (!strongSelf) {
+      return;
+    }
+
     fml::TimeDelta waitTime = fml::TimeDelta::FromMilliseconds(timeout * 1000);
     BOOL didTimeout =
-        self.shell.WaitForFirstFrame(waitTime).code() == fml::StatusCode::kDeadlineExceeded;
+        strongSelf.shell.WaitForFirstFrame(waitTime).code() == fml::StatusCode::kDeadlineExceeded;
     dispatch_async(dispatch_get_main_queue(), ^{
-      callback(didTimeout);
+      // Capture strongSelf to ensure that destruction does not occur on a background thread.
+      //
+      // The containing block, executed on a background thread, strongly captures self, then makes a
+      // blocking call to self.shell.WaitForFirstFrame(). If, during this time, all other instances
+      // of self are released, the containing block's reference would be the last one, resulting in
+      // `[FlutterEngine dealloc]` being called when it goes out of scope at the end of that block,
+      // on a background thread. FlutterEngine owns a reference to a PlatformViewsController, which
+      // owns a WeakPtrFactory whose destructor asserts that it be freed on the platform thread. To
+      // avoid this, we capture strongSelf in the current block, which is executed on the platform
+      // thread.
+      if (strongSelf) {
```

Comment:
`strongSelf` here is never nil. It's odd to put it in `if (strongSelf)`. How about just: 
```
dispatch_async(... ^{
  strongSelf;
  callback(didTimeout);
});
```
(not sure if compiler will prune it tho)

Reply:
Correct `strongSelf` is never nil. We need to capture `strongSelf` (as documented in the comment), and we need to do so in a way that avoids a compiler warning. The conditional check does that.

Reply:
The documentation on why we need to capture `strongSelf` is great - it all makes sense to me! 

Maybe also worth adding a comment above this condition, saying that `strongSelf` is never nil, but we add it here to avoid compiler warning? (also curious is it "unused variable" warning?).

Reply:
I'll update the comment. It's `Expression result unused`.

Reply:
Interesting, in swift we can do this to silence this warning: 

```
_ = strongSelf;
```

So here I'm wondering if you can simply do: 

```
strongSelf = nil;
```

Reply:
Almost certainly not, since it's a block capture, it's not assignable unless we also mark it `__block`, which I think is probably even weirder.

Reply:
Oh right right it needs to be __block.